include/deal.II/lac/petsc_ts.templates.h

```diff
@@ -284,6 +345,17 @@ namespace PETScWrappers
 
       AssertUser(user->explicit_jacobian(t, xdealii, Adealii, Pdealii),
                  "explicit_jacobian");
+      petsc_object_state_increase(P);
+
+      // Handle older versions of PETSc for which we cannot pass a MATSHELL
+      // matrix to DMSetMatType
+      if (user->need_dummy_assemble)
+        {
+          AssertTS(MatZeroEntries(P));
+          AssertTS(MatAssemblyBegin(P, MAT_FINAL_ASSEMBLY));
+          AssertTS(MatAssemblyEnd(P, MAT_FINAL_ASSEMBLY));
+          AssertTS(MatShift(P, 0.0));
```

Comment:
the reason for this is a too-restrictive check run inside the PETSc code that was requiring the diagonal entries of the matrix to be present, even if they were zero. This is because you can pass an `RHSJacobian` and then use an implicit method, that requires `J = shift * identity + RHSJacobian `, and this call would have failed if some of the diagonal entries were missing for sparse matrices because it would have triggered a reallocation of the values.